### PR TITLE
Add ChangeCenterOfStrahlkorper.

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

Adds functions for changing the expansion center of a Strahlkorper.

This is used mostly when changing Strahlkorpers from one frame to another, in which case
one often wants to reset the expansion center so that it is near the physical center.

This will also be useful for the plunge->ringdown transition of a BBH where one wishes to center a new AH on a new grid with a single excision region.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Utilities/Array.hpp
Comment (fmahebert) [inline src/Utilities/Array.hpp]: I am struggling with the same line in PR #2392. This change didn't work for me there. I think we (i.e., whoever merges first) may just have to add a NOLINT to each of these lines...

Comment (markscheel) [reply src/Utilities/Array.hpp]: My fix just moved the NOLINTNEXTLINE down a line from where it used to be.  Seems to work for me.  Clang-tidy still fails on other things though (all in a boost routine).

File: .github/workflows/Tests.yaml
Comment (geoffrey4444) [inline .github/workflows/Tests.yaml]: Just so I understand, what's the purpose of this change?

File: src/Utilities/Array.hpp
Comment (geoffrey4444) [reply src/Utilities/Array.hpp]: Clang-tidy passes for me in #2402 , where I adopted @markscheel 's approach of just moving NOLINTNEXTLINE down

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp]: Is "expansion center" standard terminology? I wonder if readers might confuse it with the expansion (as in expansion/shear) of the generators of a surface. Don't know if there's a better phrase than "expansion center" though...maybe instead say "Changes the point about which the spectral basis of the strahlkorper is expanded"?

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp:33]: In practice, how much does the tolerance change under different runs we might do? Maybe recommend or set a good default if this doesn't need to change a lot depending on the run? Does spec do something meaningful?

I ask because I'm worried about proliferation of thresholds in general

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp:25]: Is there a reason to not use a const reference, to avoid the allocation?

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: Initialize to the other component and then loop over the others

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp:48]: Can this be a const reference to avoid the allocation?

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: Should the 0.9 and 1.1 be hardcoded? If so, maybe there should be a single variable storing the 10% padding for the bracketing, so it's easy to change if needed later?

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: Should this be an option with a default?

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (geoffrey4444) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp:142]: How do you know how many iterations you'll need? I worry that this could be a point of failure in our simulations...how does spec deal with this?

File: tests/Unit/ApparentHorizons/Test_ChangeCenterOfStrahlkorper.cpp
Comment (geoffrey4444) [inline tests/Unit/ApparentHorizons/Test_ChangeCenterOfStrahlkorper.cpp]: Can't you refactor the test to have a single test function in an anonymous namespace, with a template parameter and ifs (or if constexprs) selecting change_center_of_strahlkorper_to_physical() or change_center_of_strahlkorper(), to avoid the duplicated code?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Thanks for doing this, @markscheel! I have some suggestions and questions

File: src/Utilities/Array.hpp
Comment (fmahebert) [reply src/Utilities/Array.hpp]: 👍 What I didn't realize when I wrote my earlier comment, is that I was on a slightly older `develop` where the file had different line breaks. On current develop this change does work for me 🙂

File: .github/workflows/Tests.yaml
Comment (markscheel) [reply .github/workflows/Tests.yaml]: Before this change, if the tests fail for any of the compiler options (for example "gcc-7 Debug ON") then the tests are not even run for the other options.

After this change, all the tests are run, even if some of them fail.

Not sure how important this is, but I wanted it for debugging because it wasn't clear to me why the tests were failing for some compilers but succeeding for others, and when the first compiler that was tried was a failure, I didn't even know that the tests succeeded for other compilers.

Comment (nilsvu) [reply .github/workflows/Tests.yaml]: Please make sure to drop this change before merging it into `develop`, perhaps prefix the commit message with `DROP` to remember

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp:25]: You get an allocation no matter what, since theta_phi_points() does an allocation and returns an object.

Using a const reference would work too (C++ guarantees that a reference to a temporary remains valid during the scope of that reference), but would save nothing because of return value optimization, and also would be slightly misleading.

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp:48]: I think you get an allocation no matter what, same as for theta_phi_points() above.

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp
Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp]: What if I define 'expansion center' here and then use the term later?  Then I don't need to repeat "the point about which the spectral basis of the strahlkorper is expanded" multiple times.

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp
Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp:33]: I got rid of the tolerance, made it roundoff.  It's not very expensive to do so.

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp:142]: SpEC sets maxiter to 100 and we never looked back.  But I don't think we need nearly that number of iterations. Convergence is fast because the approximation made by physical_center() is based on zeroing the l=1 modes, which makes smaller changes in the other modes that affect the center.

File: tests/Unit/ApparentHorizons/Test_ChangeCenterOfStrahlkorper.cpp
Comment (markscheel) [reply tests/Unit/ApparentHorizons/Test_ChangeCenterOfStrahlkorper.cpp]: Ok, I factored out duplicated code.

File: .github/workflows/Tests.yaml
Comment (markscheel) [reply .github/workflows/Tests.yaml]: At one of the SpECTRE meetings, it was suggested that this change be made permanent because we are not limited by the speed of failing tests.  So that's why I included it in this PR.

**Review by geoffrey4444 — COMMENTED** [review]
Happy for you to squash

File: .github/workflows/Tests.yaml
Comment (nilsvu) [reply .github/workflows/Tests.yaml]: In that case, please make it a separate PR, so it can get the attention it deserves :)

**Comment by markscheel** [comment]
Squashed.  Split off one commit into its own PR #2407.

File: .github/workflows/Tests.yaml
Comment (markscheel) [reply .github/workflows/Tests.yaml]: Ok, done. #2407

**Comment by geoffrey4444** [comment]
Looks like you still have a clang-tidy failure and some other test failures

**Comment by markscheel** [comment]
The clang-tidy error is in a boost function, not SpECTRE code, and I don't know how to suppress it.
The MacOS error I was told to ignore.
I don't understand why the other test just stopped running so I'm restarting the tests now.

**Review by geoffrey4444 — DISMISSED** [review]

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (nilsdeppe) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: this should just be an anonymous namespace since the functions in it aren't publicly consumed

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (nilsdeppe) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: Why not:
```cpp
return {sin_theta * cos(new_theta_phi[1]), sin_theta * sin(new_theta_phi[1]), cos(new_theta_phi[0])};
```
You may need more braces but I think this should work. You are able to construct vectors by their components, but I'm unsure of the expression templates will give you grief.

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (nilsdeppe) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: `const gsl::not_null` (add `const`)

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (nilsdeppe) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: add `const`

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (nilsdeppe) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: This is very strict. How sure are you that something like `2.0 * epsilon` wouldn't be better?

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp
Comment (kidder) [inline src/ApparentHorizons/ChangeCenterOfStrahlkorper.hpp]: should just need forward declarations of both of these

**Review by nilsdeppe — COMMENTED** [review]
Looks good, few minor suggestions. Please go ahead and squash anything immediately

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: Yes it is strict.  I'm happy to change it to 2.0*epsilon if you want but I have no way of justifying why it should be 2.0 versus 3.0 or 8.0.   So far just epsilon did not cause any of the random number tests to fail. Convergence is very rapid for the tests that I have done, 7 iterations at most.

Comment (nilsdeppe) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: Okay, my worry is just that it might be much easier to achieve 2 * eps simply because you are trying to get all the bits to match. I agree there isn't a reason for why a specific value times eps, but I would be more comfortable not having us trying to match something to bitwise machine precision. Having said that, it's not something I would block the PR for, just something I wanted to generate some short discussion over :)

Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: Ok, changed to 2.0*epsilon.  It will probably save an iteration in some cases.

**Comment by markscheel** [comment]
Squashed and rebased.  Clang-tidy will still fail on boost code.

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (markscheel) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp]: I didn't know you can do this!  There are similar things in my other PRs that I can change in the same way.
  Yes, there was some grief because the constructor of Tensor that takes a std::array is `explicit`, but that wasn't too hard to get around.

File: src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp
Comment (nilsdeppe) [reply src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp:25]: Ah right, it's the explicit-ness that gives issues in return types :)

**Review by nilsdeppe — APPROVED** [review]
LGTM!

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
ignoring clang-tidy error about boost root finder

**Pull request merged**
